packages/utils/crates/utils-types/src/sma.rs

```diff
@@ -1,56 +1,144 @@
     fn len(&self) -> usize;
+
+    /// Indicates whether the window is fully occupied
+    /// with samples.
+    fn is_window_full(&self) -> bool {
+        self.len() == self.window_size()
+    }
+
+    /// Indicates whether there are no samples.
+    fn is_empty(&self) -> bool;
 }
 
 /// Basic implementation of Simple Moving Average (SMA).
 /// The maximum window size is bound by 2^32 - 1.
-/// Useful mainly for floating-point types, as it does not accumulate floating point error with each sample,
-/// but requires `O(N)` of memory and `O(N)` for average computation, `N` being window size.
+/// Useful mainly for floating-point types, as it does not accumulate floating point error with each sample.
+/// Requires `O(N)` of memory and `O(N)` for average computation, `N` being window size.
+/// The divisor argument `D` is used only for such types `T` that do not implement `From<u32>` (such as `Duration`,...).
 #[derive(Clone, Debug, PartialEq)]
-pub struct NoSumSMA<T>
-where
-    T: for<'a> Add<&'a T, Output = T> + Div<T, Output = T> + Clone + From<u32> + Display + Default,
-{
+pub struct NoSumSMA<T, D = T> {
     window: AllocRingBuffer<T>,
+    _div: PhantomData<D>,
 }
 
-impl<T> SMA<T> for NoSumSMA<T>
+impl<T, D> SMA<T> for NoSumSMA<T, D>
 where
-    T: for<'a> Add<&'a T, Output = T> + Div<T, Output = T> + Clone + From<u32> + Display + Default,
+    T: for<'a> Sum<&'a T> + Div<D, Output = T>,
+    D: From<u32>,
 {
-    fn add_sample(&mut self, sample: T) {
+    fn push(&mut self, sample: T) {
         self.window.push(sample);
     }
 
-    fn get_average(&self) -> T {
-        if !self.window.is_empty() {
-            let mut ret = T::default();
-            for v in self.window.iter() {
-                ret = ret + v;
+    fn average(&self) -> Option<T> {
+        if !self.is_empty() {
+            Some(self.window.iter().sum::<T>() / D::from(self.window.len() as u32))
+        } else {
+            None
+        }
+    }
+
+    fn window_size(&self) -> usize {
+        self.window.capacity()
+    }
+
+    fn len(&self) -> usize {
+        self.window.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.window.is_empty()
+    }
+}
+
+impl<T, D> NoSumSMA<T, D>
+where
+    T: for<'a> Sum<&'a T> + Div<D, Output = T>,
+    D: From<u32>,
+{
+    /// Creates an empty SMA instance with the given window size.
+    /// The maximum window size is u32::MAX and must be greater than 1.
+    pub fn new(window_size: u32) -> Self {
+        assert!(window_size > 1, "window size must be greater than 1");
+        Self {
+            window: AllocRingBuffer::new(window_size as usize),
+            _div: PhantomData,
+        }
+    }
+
+    /// Creates SMA instance given window size and some initial samples.
+    pub fn new_with_samples(window_size: u32, initial_samples: Vec<T>) -> Self {
+        let mut ret = Self::new(window_size);
+        initial_samples.into_iter().for_each(|s| ret.push(s));
+        ret
+    }
+}
```

Comment:
There still needs to be the `window_size` passed somehow, so `From<Iterator<...>>` would not work.
But could be `initial_samples: I` where `I: IntoIter<I = T>`

Reply:
If pushed to perfection, the window size does not change, so it could be a genrric usize argument.

Reply:
It cannot unfortunately, since we need to be able to change it at runtime. This was the reason we abandoned the 3rd party crate (which had it as const generic) and implement it ourselves